pox/pox/ext/jellyfish_controller.py

```diff
@@ -0,0 +1,143 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at:
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+JellyfishController class.
+"""
+
+from pox.core import core
+from pox.lib.util import dpidToStr
+from utils import build_topology, dpid_to_str
+from pox.lib.revent import EventMixin
+import pox.openflow.libopenflow_01 as of
+from routing import Routing
+from switch import Switch
+
+log = core.getLogger()
+
+class JellyfishController (EventMixin):
+  """
+  An JellyfishController object is created once, and it is in charge of instantiating
+  all of the switches in the Controller as they come up in the mininet topology.
+
+  A JellyfishController subclasses EventMixin, because it is triggered when
+  certain events happen in the mininet network.
+
+  The Jellyfish Controller consumes ALL packet_in events and demultiplexes
+  them to the correct switch and switch action, after determining what
+  action must occur using an internal routing class.
+  """
+  def __init__ (self, topology, routing):
+    # Keep track of the connection to the switch so that we can
+    # send it messages!
+    self.routing = routing
+    for i in routing.rtable.items():
+      log.info(i)
+
+    self.switches = {}  # Switches seen: [dpid] -> Switch
+    self.topology = topology  # Master Topo object, passed in and never modified.
+    self.routing = routing  # Master Routing object, passed in and reused.
+    self.all_switches_up = False  # Sequences event handling.
+
+    # Make this controller listen to openflow events, like when switches come up
+    # or when a packet comes into a switch.
+    self.listenTo(core.openflow, priority=0)
+
+  def forward(self, packet, switch, egress_port):
+    """
+    Forward a packet along the given egress port of
+    the given switch.
+    """
+    log.info("Sending packet out of port %d from switch %d" % (egress_port, switch.dpid))
+    switch.send_packet_data(egress_port, packet)
+
+  def _handle_ConnectionUp (self, event):
+    """
+    Is called whenever a switch in the Mininet topoplogy comes up,
+    and registers it in the controller.
+    """
+
+    switch_dpid = event.dpid
+    switch = self.switches.get(event.dpid)
+
+    # The name of the switch as known by the topology object.
+    switch_name_str = dpid_to_str(switch_dpid)
+
+    # Controller ignores switch if this is not a switch we recognize.
+    if switch_name_str not in self.topology.switches():
+      log.warn("Ignoring unknown switch %s" % switch_name_str)
+      return
+
+    # We expect the switch to be None, so we create it.
+    if switch is None:

```

Comment:
Don't we also need to add this to routing?

Reply:
Yeah, the routing class needs to probably do some "registering" of the switch that came up internally.  I'll add a dummy method to routing that ingests this switch object